syntax = "proto3";

package modelservice;

service ModelService {
  rpc UploadData (UploadDataRequest) returns (UploadDataResponse);
  rpc TrainModel (TrainModelRequest) returns (TrainModelResponse);
  rpc ListAvailableModels (Empty) returns (ListAvailableModelsResponse);
  rpc Predict (PredictRequest) returns (PredictResponse);
  rpc DeleteModel (DeleteModelRequest) returns (DeleteModelResponse);
  rpc Status (Empty) returns (StatusResponse);
  rpc ListTrainedModels (Empty) returns (ListTrainedModelsResponse);
  rpc GetPredictions (GetPredictionsRequest) returns (GetPredictionsResponse);
  rpc UpdateModel (UpdateModelRequest) returns (UpdateModelResponse);
}

message UploadDataRequest {
  string data = 1;
}

message UploadDataResponse {
  string message = 1;
}

message TrainModelRequest {
  string model_type = 1;
  map<string, int64> hyperparameters = 2;
  string target_variable = 3;
}

message TrainModelResponse {
  string message = 1;
  string model_id = 2;
}

message Empty {}

message ListAvailableModelsResponse {
  repeated string model_types = 1;
}

message PredictRequest {
  string model_id = 1;
  string data = 2;
}

message PredictResponse {
  repeated string prediction = 1;
}

message DeleteModelRequest {
  string model_id = 1;
}

message DeleteModelResponse {
  string message = 1;
}

message StatusResponse {
    string status = 1;
}

message ListTrainedModelsResponse {
  repeated string model_ids = 1;
}

message GetPredictionsRequest {
  string model_id = 1;
}

message GetPredictionsResponse {
  repeated string predictions = 1;
}

message UpdateModelRequest {
  string model_id = 1;
  map<string, int64> hyperparameters = 2;
  string target_variable = 3;
}

message UpdateModelResponse {
  string message = 1;
}
